kotlinx-coroutines-debug/src/internal/DebugProbesImpl.kt

```diff
@@ -117,22 +117,92 @@ internal object DebugProbesImpl {
                     "${value.state} (Last suspension stacktrace, not an actual stacktrace)"
                 else
                     value.state.toString()
 
                 append("\n\nCoroutine $key, state: $state")
-                val observedStackTrace = value.lastObservedStackTrace()
                 if (observedStackTrace.isEmpty()) {
                     append("\n\tat ${artificialFrame(ARTIFICIAL_FRAME_MESSAGE)}")
                     printStackTrace(value.creationStackTrace)
                 } else {
-                    printStackTrace(value.lastObservedStackTrace())
+                    printStackTrace(enhancedStackTrace)
                 }
             }
     }
 
+    /**
+     * Tries to enhance [coroutineTrace] (obtained by call to [CoroutineState.lastObservedStackTrace]) with
+     * thread dump of [CoroutineState.lastObservedThread].
+     *
+     * Returns [coroutineTrace] if enhancement was unsuccessful or the enhancement result.
+     */
+    private fun enhanceStackTraceWithThreadDump(
+        state: CoroutineState,
+        coroutineTrace: List<StackTraceElement>
+    ): List<StackTraceElement> {
+        val coroutineStackTrace = coroutineTrace as ArrayList<StackTraceElement>
+        val thread = state.lastObservedThread
+        if (state.state != State.RUNNING || thread == null) return coroutineStackTrace
+        // Avoid security manager issues
+        val actualTrace = kotlin.runCatching { thread.stackTrace }.getOrNull()
+            ?: return coroutineStackTrace
+
+        /*
+         * Here goes heuristic that tries to merge two stacktraces: real one
+         * (that has at least one but usually not so many suspend function frames)
+         * and coroutine one that has only suspend function frames.
+         *
+         * Heuristic:
+         * 1) Dump lastObservedThread
+         * 2) Find the next frame after BaseContinuationImpl.resumeWith (continuation machinery).
+         *   Invariant: this method is called under the lock, so such method **should** be present
+         *   in continuation stacktrace.
+         *
+         * 3) Find target method in continuation stacktrace (metadata-based)
+         * 4) Prepend dumped stacktrace (trimmed by target frame) to continuation stacktrace
+         *
+         * Heuristic may fail on recursion and overloads, but it will be automatically improved
+         * with KT-29997
+         */
+        val indexOfResumeWith = actualTrace.indexOfFirst {
+            it.className == "kotlin.coroutines.jvm.internal.BaseContinuationImpl" &&
+                    it.methodName == "resumeWith" &&
+                    it.fileName == "ContinuationImpl.kt"
+        }
+
+        // We haven't found "BaseContinuationImpl.resumeWith" resume call in stacktrace
+        // This is some inconsistency in machinery, do not fail, fallback
+        if (indexOfResumeWith < 1) {
+            return coroutineStackTrace
+        }
```

Comment:
A minor note on style: this kinds of of "early return ifs" a better written in single-line. Consider that, but I would not insist for this particular one. You can also make more explicit alignment of this check if the code on the next line after it (blow) by dropping this whole `if` and rewriting the like below:
```
val continuationFrame = actualTrace.getOrNull(indexOfResumeWith - 1) 
    ?: return coroutineStackTrace
```
The advantage is that reader will not have to think if the check was correct and whether this line can fail or not.